// tsparticles/dist/Utils/Plugins.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Plugins = void 0;
class Plugins {
    static getPlugin(plugin) {
        return Plugins.plugins.filter((t) => t.id === plugin)[0];
    }
    static addPlugin(plugin) {
        if (!Plugins.getPlugin(plugin.id)) {
            Plugins.plugins.push(plugin);
        }
    }
    static getAvailablePlugins(container) {
        const res = new Map();
        const availablePlugins = Plugins.plugins.filter((t) => t.needsPlugin(container.options));
        for (const plugin of availablePlugins) {
            res.set(plugin.id, plugin.getPlugin(container));
        }
        return res;
    }
    static loadOptions(options, sourceOptions) {
        for (const plugin of Plugins.plugins) {
            plugin.loadOptions(options, sourceOptions);
        }
    }
    static getPreset(preset) {
        return Plugins.presets.get(preset);
    }
    static addPreset(presetKey, options) {
        if (!Plugins.getPreset(presetKey)) {
            Plugins.presets.set(presetKey, options);
        }
    }
    static addShapeDrawer(type, drawer) {
        if (!Plugins.getShapeDrawer(type)) {
            Plugins.drawers.set(type, drawer);
        }
    }
    static getShapeDrawer(type) {
        return Plugins.drawers.get(type);
    }
    static getSupportedShapes() {
        return Plugins.drawers.keys();
    }
}
exports.Plugins = Plugins;
Plugins.plugins = [];
Plugins.presets = new Map();
Plugins.drawers = new Map();
